main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -371,11 +371,23 @@ object Lowering {
       val t = visitType(tpe)
       LoweredAst.Expr.Lambda(p, e, t, loc)
 
-    case TypedAst.Expr.ApplyClo(exp1, exp2, tpe, eff, loc) =>
-      val e1 = visitExp(exp1)
-      val e2 = visitExp(exp2)
+    case TypedAst.Expr.ApplyClo(exp, exps, argEffs, tpe, eff, loc) =>
+      val e = visitExp(exp)
+      val es = exps.map(visitExp)
       val t = visitType(tpe)
-      LoweredAst.Expr.ApplyClo(e1, e2, t, eff, loc)
+      // Curried function of `exps` arguments.
+      val funType = exps.zip(argEffs).foldRight(tpe){
+        case ((e, argEff), res) => Type.mkArrowWithEffect(e.tpe, argEff, res, loc.asSynthetic)
+      }
+      val (applies, _) = es.foldLeft((e, funType)){
+        case ((f, Type.Apply(Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Arrow(2), _), argEff, _), _argT, _), next, _)), arg) =>
+          (
+            LoweredAst.Expr.ApplyClo(f, arg, next, Type.mkUnion(argEff, arg.eff, loc.asSynthetic), loc.asSynthetic),
+            next
+          )
+        case ((_, tpe), _) => throw InternalCompilerException(s"Unexpected type: $tpe", loc)
+      }
+      applies
```

Comment:
I need the argEffs in lowering to split the application into n things again - we have a similar kind of thing with applyDef and monomorph - sometimes we need extra type info for backend transformations

Reply:
Can't we do this using the type of exp?

Reply:
argEffs might be a misleading name, its the effect of the part of the arrow related to that one argument

Reply:
But the arrow holds this information, right?

Reply:
It does yes - it is just scary to me to destruct types but it is also done other places in lowering so maybe its okay

Reply:
edit: it is not okay